modules/src/ics04_channel/handler/chan_open_ack.rs

```diff
@@ -0,0 +1,353 @@
+//! Protocol logic specific to ICS4 messages of type `MsgChannelOpenAck`.
+use crate::events::IBCEvent;
+use crate::handler::{HandlerOutput, HandlerResult};
+use crate::ics03_connection::connection::State as ConnectionState;
+use crate::ics04_channel::channel::{ChannelEnd, Counterparty, State};
+use crate::ics04_channel::context::ChannelReader;
+use crate::ics04_channel::error::{Error, Kind};
+use crate::ics04_channel::events::Attributes;
+use crate::ics04_channel::handler::verify::verify_proofs;
+use crate::ics04_channel::handler::ChannelResult;
+use crate::ics04_channel::msgs::chan_open_ack::MsgChannelOpenAck;
+use Kind::ConnectionNotOpen;
+
+pub(crate) fn process(
+    ctx: &dyn ChannelReader,
+    msg: MsgChannelOpenAck,
+) -> HandlerResult<ChannelResult, Error> {
+    let mut output = HandlerOutput::builder();
+
+    // Unwrap the old channel end and validate it against the message.
+
+    let mut channel_end = ctx
+        .channel_end(&(msg.port_id().clone(), msg.channel_id().clone()))
+        .ok_or_else(|| Kind::ChannelNotFound.context(msg.channel_id().clone().to_string()))?;
+
+    // Validate that the channel end is in a state where it can be ack.
+
+    if !channel_end.state_matches(&State::Init) && !channel_end.state_matches(&State::TryOpen) {
+        return Err(Into::<Error>::into(Kind::InvalidChannelState(
+            msg.channel_id().clone(),
+        )));
+    }
+    //?? Why don't we check that the  channel_end's version matches
+    // the counterparty_version received in the message.
```

Comment:
I believe this would be checked in the application callback (which we currently don't have)

Reply:
Would it hurt to add a check here too ?

Reply:
The problem is the handler doesn't know how to check version compatibility at channel level. The two strings might be different but still be compatible.

Reply:
ok